### ISSUE TITLE
TuringGLM sampling quite slow compared to brms (rstan) and bambi (pymc)

### ISSUE THREAD

**Issue by sreedta8** [TuringLang/TuringGLM.jl#57]
Hi

I'm running a hierarchical model using ```TuringGLM```. I had previously run the same model using ```brms (rstan)``` and ```bambi (pymc)```. This data set consists of 12 months of prescribing from 1000 physicians. The specific columns include the number of new Rx they have written for a drug, along with adstocks computed for each physician on six marketing variables.

 Model tested: ```nrx ~ calls80 +samples10 + lle50 +spe80 + copay10 + fto10 + (calls80 | mdmidc) ```
6 population level fixed effects and 1 random effects for the group variable ```mdmidc``` (at an individual HCP level random effects)

Command used to sample is ```chn = sample(model, NUTS(), 3_000, progress=true)```
1. How do I specify ```target acceptance``` in the above command?
2. How do I specify ```adaptation steps``` in the above command?

Here are the specifications:
1. Intel based Laptop running Ubuntu 21.04 with 8GB of RAM
3. warmup / adaptation steps = 1000
4. number of samples (post warmup) = 4000
5. target acceptance = 0.98

Time to complete the sampling process:
1. brms = ~40 minutes (range of 38 minutes to 43 minutes over 10 runs)
2. bambi = ~41 minutes (range of 37 minutes to 45 minutes over 10 runs)
3. TuringGLM - Sampling has not yet started after 45 minutes - I keep seeing these error messages

```
┌ Warning: The current proposal will be rejected due to numerical error(s).
│   isfinite.((θ, r, ℓπ, ℓκ)) = (true, false, false, false)
└ @ AdvancedHMC /home/sreedta/.julia/packages/AdvancedHMC/51xgc/src/hamiltonian.jl:47
┌ Warning: The current proposal will be rejected due to numerical error(s).
│   isfinite.((θ, r, ℓπ, ℓκ)) = (true, false, false, false)
└ @ AdvancedHMC /home/sreedta/.julia/packages/AdvancedHMC/51xgc/src/hamiltonian.jl:47
```

**Comment by sreedta8** [comment]
After about 1 hour and 15 minutes I see a message about ```step size```.

```Info: Found initial step size
│   ϵ = 9.765625e-5
└ @ Turing.Inference /home/sreedta/.julia/packages/Turing/GxgQ1/src/inference/hmc.jl:188 ```

I feel as if the default values I'm using for ```number of adaptations``` and ```target acceptance``` are impeding this model.

I would be happy to share my data set if anyone else wants to try the hierarchical model I'm testing.

**Comment by sreedta8** [comment]
I stopped the original sampling command and restarted the process again. This time the initial step size was found quickly. However the ETA (see below) to complete the sampling is *2 days, 6:29:51*. I have no doubt I'm doing something awfully wrong.

```
Warning: The current proposal will be rejected due to numerical error(s).
│   isfinite.((θ, r, ℓπ, ℓκ)) = (true, false, false, false)
└ @ AdvancedHMC /home/sreedta/.julia/packages/AdvancedHMC/51xgc/src/hamiltonian.jl:47
┌ Warning: The current proposal will be rejected due to numerical error(s).
│   isfinite.((θ, r, ℓπ, ℓκ)) = (true, false, false, false)
└ @ AdvancedHMC /home/sreedta/.julia/packages/AdvancedHMC/51xgc/src/hamiltonian.jl:47
┌ Warning: The current proposal will be rejected due to numerical error(s).
│   isfinite.((θ, r, ℓπ, ℓκ)) = (true, false, false, false)
└ @ AdvancedHMC /home/sreedta/.julia/packages/AdvancedHMC/51xgc/src/hamiltonian.jl:47
┌ Warning: The current proposal will be rejected due to numerical error(s).
│   isfinite.((θ, r, ℓπ, ℓκ)) = (true, false, false, false)
└ @ AdvancedHMC /home/sreedta/.julia/packages/AdvancedHMC/51xgc/src/hamiltonian.jl:47
┌ Warning: The current proposal will be rejected due to numerical error(s).
│   isfinite.((θ, r, ℓπ, ℓκ)) = (true, false, false, false)
└ @ AdvancedHMC /home/sreedta/.julia/packages/AdvancedHMC/51xgc/src/hamiltonian.jl:47
┌ Info: Found initial step size
│   ϵ = 0.000390625
└ @ Turing.Inference /home/sreedta/.julia/packages/Turing/GxgQ1/src/inference/hmc.jl:188
Sampling:   2%|▋                                        |  ETA: 2 days, 6:29:51
```

**Comment by sreedta8** [comment]
I figured out how to specify ```number of adaptations``` and ```target acceptance``` using

```chn = sample(model, NUTS(1000, 0.98), 3_000, progress=true);```

With this specification, the sampling ETA is only a little better:

```
Info: Found initial step size
│   ϵ = 0.000390625
└ @ Turing.Inference /home/sreedta/.julia/packages/Turing/GxgQ1/src/inference/hmc.jl:188
Sampling:   4%|█▊                                       |  ETA: 1 days, 12:59:29
```

Would love some feedback on how I can improve the posterior sampling time to be in line with ```brms``` and ```bambi```

**Comment by storopoli** [comment]
@sreedta8 `TuringGLM.jl` currently only supports single random-intercept models, i.e. `(1 | grouping_variable)`.

Have you tried coding it straight up using Turing?
You could do something like:

```julia
@model function varying_slope_ncp(X, idx, y, calls80; n_gr=length(unique(idx)), predictors=size(X, 2))
    # priors
    α ~ Normal(mean(y), 2.5 * std(y))       # population-level intercept
    β ~ filldist(Normal(0, 2), predictors)  # population-level coefficients
    σ ~ Exponential(1 / std(y))             # residual SD

    # prior for variance of random slopes
    # usually requires thoughtful specification
    τ ~ truncated(Cauchy(0, 2), 0, Inf)    # group-level SDs slopes
    zⱼ ~ filldist(Normal(0, 1), n_gr)      # NCP group-level slopes

    # likelihood
    ŷ = α .+ X * β .+ (zⱼ[idx] .* τ .* calls80)
    y ~ MvNormal(ŷ, σ)
end;
```

This is the non-centered parameterization.

where:

- `X`: is a matrix for all variables except `nrx` and `mdmidc`, i.e. `Matrix(select(df, Not([:nrx, :mdmidc])))`
- `idx`: a vector of ids for the grouping variable in Integer values, i.e. `df[:, :mdmidc_omt]`. You can convert them to `Int64` using `CategoricalArrays.jl`'s `levelcode` function.
- `y`: a vector of values from the dependent variable `df[:, :nrx]`
- `calls80`: a vector of values from the `calls80` variable `df[:, :calls80]`

Take a look at this tutorial: https://storopoli.github.io/Bayesian-Julia/pages/10_multilevel_models

**Comment by sreedta8** [comment]
@storopoli thank you so much for your feedback. So far I have only tried one random intercept via TuringGLM using ```(1 | mdmidc)```. I will review the link and try a direct specification in Turing and let you know of the results.

**Comment by sreedta8** [comment]
@storopoli I have specified the following below based on your input and the example models. I got the initial step size quickly. But after 2.5 hours, I still do not see the sampling progress bar (I set ```Turing.setprogress!(true);```)

```
@model function varying_slope(X, idx, y, calls80; n_gr=length(unique(idx)), predictors=size(X, 6))
    # priors
    α ~ Normal(mean(y), 2.5 * std(y))       # population-level intercept
    β ~ filldist(Normal(0, 2), predictors)  # population-level coefficients
    σ ~ Exponential(1 / std(y))             # residual SD

    # prior for variance of random slopes
    # usually requires thoughtful specification
    τ ~ truncated(Cauchy(0, 2), 0, Inf)    # group-level SDs slopes
    zⱼ ~ filldist(Normal(0, 1), n_gr)      # NCP group-level slopes

    # likelihood
    ŷ = α .+ X * β .+ (zⱼ[idx] .* τ .* calls80)
    y ~ MvNormal(ŷ, σ)
end;

X = Matrix(select(hcp, ([:calls80, :samples10, :lle50, :spe80, :copay10, :fto10]))) # matrix of predictors
idx = hcp[:, :mdmid]
y = hcp[:, :nrx]
calls80 = hcp[:, :calls80]

Turing.setprogress!(true);
model_slope = varying_slope(X, idx, y, calls80; n_gr=length(unique(idx)), predictors=size(X, 2))
chain_slope = sample(model_slope, NUTS(1_000, 0.98), MCMCThreads(), 3_000, 2)
```
Output:

┌ Info: [Turing]: progress logging is enabled globally
└ @ Turing /home/sreedta/.julia/packages/Turing/GxgQ1/src/Turing.jl:22
┌ Info: [AdvancedVI]: global PROGRESS is set as true
└ @ AdvancedVI /home/sreedta/.julia/packages/AdvancedVI/W2zsz/src/AdvancedVI.jl:15
┌ Warning: Only a single thread available: MCMC chains are not sampled in parallel
└ @ AbstractMCMC /home/sreedta/.julia/packages/AbstractMCMC/fnRmh/src/sample.jl:291
┌ Warning: The current proposal will be rejected due to numerical error(s).
│   isfinite.((θ, r, ℓπ, ℓκ)) = (true, false, false, false)
└ @ AdvancedHMC /home/sreedta/.julia/packages/AdvancedHMC/51xgc/src/hamiltonian.jl:47
┌ Warning: The current proposal will be rejected due to numerical error(s).
│   isfinite.((θ, r, ℓπ, ℓκ)) = (true, false, false, false)
└ @ AdvancedHMC /home/sreedta/.julia/packages/AdvancedHMC/51xgc/src/hamiltonian.jl:47
┌ Warning: The current proposal will be rejected due to numerical error(s).
│   isfinite.((θ, r, ℓπ, ℓκ)) = (true, false, false, false)
└ @ AdvancedHMC /home/sreedta/.julia/packages/AdvancedHMC/51xgc/src/hamiltonian.jl:47
┌ Warning: The current proposal will be rejected due to numerical error(s).
│   isfinite.((θ, r, ℓπ, ℓκ)) = (true, false, false, false)
└ @ AdvancedHMC /home/sreedta/.julia/packages/AdvancedHMC/51xgc/src/hamiltonian.jl:47
┌ Warning: The current proposal will be rejected due to numerical error(s).
│   isfinite.((θ, r, ℓπ, ℓκ)) = (true, false, false, false)
└ @ AdvancedHMC /home/sreedta/.julia/packages/AdvancedHMC/51xgc/src/hamiltonian.jl:47
┌ Info: Found initial step size
│   ϵ = 0.000390625
└ @ Turing.Inference /home/sreedta/.julia/packages/Turing/GxgQ1/src/inference/hmc.jl:188
```

**Comment by ChristianMichelsen** [comment]
If you run with more than 1 thread, the progress only shows the fraction of threads that are fully done (which I personally would like to be changed, but that's another issue). 

Try running it again but with: 
```
sample(model_slope, NUTS(1_000, 0.98), 3_000)
```

This should allow you to get a better time estimate. When you have that, and is satisfied with it, I would change it back to the multi-threaded version.

**Comment by sreedta8** [comment]
@ChristianMichelsen thanks for the tip. I have tried that and the process still hangs. The sampling is not being completed. So far the only successful sampling I had was when I used TuringGLM, but it took 10.5 hours in total.

**Comment by storopoli** [comment]
I am transferring this issue to `Turing.jl` since this is a feature that `TuringGLM.jl` does not support and the user is having issues with a raw `Turing.jl` model and not one generated by `TuringGLM.jl`.
